### DISCUSSION TITLE
adding-localisation-to-an-asp-net-core-application/

### DISCUSSION THREAD

**Discussion by andrewlock** [andrewlock/blog-comments#166] [Announcements]
# adding-localisation-to-an-asp-net-core-application/

In this post I'll walk through the process of adding localisation to an ASP.NET Core application using the recommended approach with resx resource files.

https://andrewlock.net/adding-localisation-to-an-asp-net-core-application/

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at September 22 2016, 05:51, in reply to [Sławomir Bryś](https://disqus.com/by/slawomirbrys/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/166#discussioncomment-6005407)</em>

---

<p>Great question, it's something I have wanted to do as well. You can achieve something similar to this by using a SharedResource as the type parameter to an IStringLocalizer (see <a href="https://docs.asp.net/en/latest/fundamentals/localization.html#make-the-app-s-content-localizable" rel="nofollow noopener" title="https://docs.asp.net/en/latest/fundamentals/localization.html#make-the-app-s-content-localizable">the docs</a> for details) but I don't think there's a way of achieving it fully with this approach. Using this technique excludes the IViewLocalizer and DataAnnotation localisation so they would all still be in separate files.</p><p>I think the only way to achieve exactly what you want is to use a custom IStringLocalizerFactory and IStringLocalizer. I've started to play with this for a follow up post but it's a bit laborious. Hopefully I'm missing something and there's an easy toggle option somewhere ;)</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at September 25 2016, 10:52, in reply to [hishamco_2007](https://disqus.com/by/hishamco_2007/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/166#discussioncomment-6005413)</em>

---

<p>Thanks Hisham, I've updated the link. That's a really nice culture switcher in your post, good work -  I've added your links to the post :)</p><p>The problem I had with the config in the docs is that the RequestLocalizationOptions  are created and passed to the middleware in the Configure method, instead of added in the ConfigureServices method and then obtained from the Services in Configure. With the documented approach, the options didn't seem to be used correctly.</p><p>I haven't dug into why it doesn't work, and to be fair, I haven't actually checked out the Localization Sample code project that's referenced in the documentation to see if it works correctly in that case, but when I get round to it I will definitely send a PR:)</p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at September 28 2016, 09:43, in reply to [hishamco_2007](https://disqus.com/by/hishamco_2007/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/166#discussioncomment-6005417)</em>

---

<p>Hi Hisham, yeah, it was configuring RequestLocalizationOptions using version 1 which did not appear to work for me.</p><p>In fact, I've just twigged - it didn't work because I was injecting the RequestLocalizationOptions into the layout for the culture picker. With approach number 1, the Options are not registered with the DI container, so you can't use that approach, and the configured cultures are not available to the picker</p><p>So, my mistake, but probably a good idea to make that clearer in the docs if you will need access to the supported culture anywhere else (e.g. culture picker). I'll update the post with the details shortly. Cheers!</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at January 12 2017, 10:05, in reply to [João Simões](https://disqus.com/by/gravity00/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/166#discussioncomment-6005422)</em>

---

<p>Hi, I have written a series of posts on this approach now - the last one here (which lists all the articles) is here <a href="http://andrewlock.net/redirecting-unknown-cultures-to-the-default-culture-when-using-the-url-culture-provider/" rel="nofollow noopener" title="http://andrewlock.net/redirecting-unknown-cultures-to-the-default-culture-when-using-the-url-culture-provider/">http://andrewlock.net/redirecting-unknown-cultures-to-the-default-culture-when-using-the-url-culture-provider/</a></p><p>Hope it helps!:)</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at January 12 2017, 10:04, in reply to [Oliver Waterkamp](https://disqus.com/by/oliverwaterkamp/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/166#discussioncomment-6005425)</em>

---

<p>Thanks, appreciate the kind words:)</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 08 2017, 08:25, in reply to [Stefano Baruzzo](https://disqus.com/by/Stefanone_91/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/166#discussioncomment-6005431)</em>

---

<p>Thanks, nice spot, I'll update the code, thanks!</p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at September 01 2017, 08:02, in reply to [Stefano Baruzzo](https://disqus.com/by/Stefanone_91/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/166#discussioncomment-6005433)</em>

---

<p>HI, that's strange that you can't get it to work. Looking at the source code for FileExtensionsAttribute, it explicitly says it should be localizable (and it inherits from ValidationAttribute, so should be) <a href="https://github.com/dotnet/corefx/blob/master/src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/FileExtensionsAttribute.cs" rel="nofollow noopener" title="https://github.com/dotnet/corefx/blob/master/src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/FileExtensionsAttribute.cs">https://github.com/dotnet/corefx/blob/master/src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/FileExtensionsAttribute.cs</a> If this definitely doesn't work, I suggest putting together a minimal reproduction sample and raising an issue on the corefx repository</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at March 14 2017, 07:35, in reply to [Deepak Khopade](https://disqus.com/by/deepakkhopade/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/166#discussioncomment-6005439)</em>

---

<p>When you say its not working, in what way? What errors are you getting?</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at May 25 2017, 08:58, in reply to [Massimo Giambona](https://disqus.com/by/massimogiambona/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/166#discussioncomment-6005448)</em>

---

<p>Thanks, fixed</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at August 21 2017, 07:15, in reply to [Dmitry Sikorsky](https://disqus.com/by/dmitrysikorsky/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/166#discussioncomment-6005454)</em>

---

<p>That isn't something I've looked at I'm afraid, but I'd be interested to know if you managed to solve it?</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at October 11 2017, 09:48, in reply to [Rabadash8820](https://disqus.com/by/rabadash8820/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/166#discussioncomment-6005462)</em>

---

<p>Hi, thanks for reading.</p><p>1) The order you call UseStaticFiles() and UseRequestLocalization() determines the order the middleware is added to the pipeline. So a request passes through each middleware in turn. If the static file middleware returns a response, then no later middleware will execute. I placed Static files before request localization as you don't need localization to serve the static files.</p><p>2)  You don't have to set resources path, as it defaults to Resources, I just added it to be explicit!</p><p>Hope that helps</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 17 2019, 05:37, in reply to [Konrad Viltersten](https://disqus.com/by/konradviltersten/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/166#discussioncomment-6005471)</em>

---

<p>Sure, that's outside the scope of this post. Damien shows an approach you can use here: <a href="https://damienbod.com/2017/11/01/shared-localization-in-asp-net-core-mvc/" rel="nofollow noopener" title="https://damienbod.com/2017/11/01/shared-localization-in-asp-net-core-mvc/">https://damienbod.com/2017/11/01/shared-localization-in-asp-net-core-mvc/</a></p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at April 30 2020, 06:33, in reply to [dush88c](https://disqus.com/by/disqus_5di5iot566/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/166#discussioncomment-6005478)</em>

---

<p>Hi, I have a suspicion you're problem is the same as I had in this post: <a href="https://andrewlock.net/dotnet-core-docker-and-cultures-solving-culture-issues-porting-a-net-core-app-from-windows-to-linux/" rel="nofollow noopener" title="https://andrewlock.net/dotnet-core-docker-and-cultures-solving-culture-issues-porting-a-net-core-app-from-windows-to-linux/">https://andrewlock.net/dotnet-core-docker-and-cultures-solving-culture-issues-porting-a-net-core-app-from-windows-to-linux/</a></p><p>Hope that helps!</p>

[a comment by an automated account is not shown]

**Comment by ejlalzadeh** [comment]
Hi Andrew, Thank you for your great article. I did not find your next article, which was related to the problems of this method and the magic strings. Can you guide me?
